File: dashboard/blueprints/standings.py
from flask import Blueprint, render_template, jsonify, make_response
from data.season_2026 import get_driver_standings, get_constructor_standings
from data.jolpica_client import JolpicaClient
import time

standings_bp = Blueprint('standings', __name__)

# Simple in-memory TTL cache for live standings (60s) — massively faster than Jolpica round-trip
_CACHE = {}
_CACHE_TTL = 60
def _get_cached(key):
    v = _CACHE.get(key)
    if v and time.time() - v[1] < _CACHE_TTL:
        return v[0]
    return None
def _set_cached(key, val):
    _CACHE[key] = (val, time.time())

def _normalize_driver_standings(payload):
    """Normalize Jolpica MRData to [{position, driver_code, points, team}] for UI."""
    try:
        data = payload.get('data') or {}
        mr = data.get('MRData') if isinstance(data, dict) else None
        if mr:
            lst = mr['StandingsTable']['StandingsLists'][0]['DriverStandings']
            team_map = {"mercedes":"mercedes","ferrari":"ferrari","mclaren":"mclaren","red_bull":"redbull","rb":"racingbulls","alpine":"alpine","haas":"haas","audi":"audi","williams":"williams","aston_martin":"astonmartin","cadillac":"cadillac"}
            out = []
            for d in lst:
                out.append({"position": int(d["position"]), "driver_code": d["Driver"]["code"], "points": int(d["points"]), "wins": int(d.get("wins",0)), "team": team_map.get(d["Constructors"][0]["constructorId"], d["Constructors"][0]["constructorId"])})
            return out
    except Exception:
        pass
    return None

def _normalize_constructor_standings(payload):
    try:
        data = payload.get('data') or {}
        mr = data.get('MRData') if isinstance(data, dict) else None
        if mr:
            lst = mr['StandingsTable']['StandingsLists'][0]['ConstructorStandings']
            cmap = {"red_bull":"redbull","rb":"racingbulls","aston_martin":"astonmartin"}
            out=[]
            for d in lst:
                cid=d["Constructor"]["constructorId"]
                out.append({"position": int(d["position"]), "team_id": cmap.get(cid,cid), "points": int(d["points"]), "wins": int(d.get("wins",0))})
            return out
    except Exception:
        pass
    return None

@standings_bp.route('/')
def index():
    return render_template('standings.html')

@standings_bp.route('/api/driver-standings')
def api_driver_standings():
    """Live driver standings — TTL cached 60s, ~2ms vs 800ms live."""
    cached = _get_cached('driver')
    if cached:
        resp = make_response(jsonify(cached))
        resp.headers['Cache-Control'] = 'public, max-age=30'
        resp.headers['X-Cache'] = 'HIT'
        return resp
    try:
        client = JolpicaClient()
        from config.settings import settings
        result = client.get_driver_standings(settings.SEASON_YEAR)
        if result.get('source') in ('live','cached'):
            norm = _normalize_driver_standings(result)
            if norm:
                payload = {'data': norm, 'source': result.get('source'), 'provenance': result.get('provenance')}
                _set_cached('driver', payload)
                resp = make_response(jsonify(payload))
                resp.headers['Cache-Control'] = 'public, max-age=30'
                resp.headers['X-Cache'] = 'MISS'
                return resp
        standings = get_driver_standings()
        payload = {'data': standings, 'source': 'local'}
        _set_cached('driver', payload)
        resp = make_response(jsonify(payload))
        resp.headers['Cache-Control'] = 'public, max-age=30'
        return resp
    except Exception as e:
        standings = get_driver_standings()
        payload = {'data': standings, 'source': 'local', 'error': str(e)}
        return jsonify(payload)

@standings_bp.route('/api/constructor-standings')
def api_constructor_standings():
    """Live constructor standings — TTL cached 60s."""
    cached = _get_cached('constructor')
    if cached:
        resp = make_response(jsonify(cached))
        resp.headers['Cache-Control'] = 'public, max-age=30'
        resp.headers['X-Cache'] = 'HIT'
        return resp
    try:
        client = JolpicaClient()
        result = client.get_constructor_standings()
        if result.get('source') in ('live','cached'):
            norm = _normalize_constructor_standings(result)
            if norm:
                payload = {'data': norm, 'source': result.get('source'), 'provenance': result.get('provenance')}
                _set_cached('constructor', payload)
                resp = make_response(jsonify(payload))
                resp.headers['Cache-Control'] = 'public, max-age=30'
                resp.headers['X-Cache'] = 'MISS'
                return resp
        standings = get_constructor_standings()
        payload = {'data': standings, 'source': 'local'}
        _set_cached('constructor', payload)
        resp = make_response(jsonify(payload))
        resp.headers['Cache-Control'] = 'public, max-age=30'
        return resp
    except Exception as e:
        standings = get_constructor_standings()
        return jsonify({'data': standings, 'source': 'local', 'error': str(e)})

File: dashboard/static/js/standings.js
/**
 * standings.js — real driver/constructor standings.
 * Handles both API shapes: {data:[...], source:'local'|'simulated'} from the
 * fallback path, and a best-effort parse of the raw Jolpica/Ergast shape if
 * the live path ever returns 'live' (see jolpica_client.py get()).
 */
(function () {
  "use strict";
  const $ = (s) => document.querySelector(s);

  function normalizeDrivers(payload) {
    if (payload.source === "live" && payload.MRData) {
      try {
        const list = payload.MRData.StandingsTable.StandingsLists[0].DriverStandings;
        return list.map((e) => ({
          position: Number(e.position),
          driver_code: e.Driver.code || (e.Driver.familyName || "").slice(0, 3).toUpperCase(),
          points: Number(e.points),
          team: (e.Constructors && e.Constructors[0] && e.Constructors[0].constructorId) || "",
        }));
      } catch (e) { /* fall through */ }
    }
    return payload.data || [];
  }
  function normalizeConstructors(payload) {
    if (payload.source === "live" && payload.MRData) {
      try {
        const list = payload.MRData.StandingsTable.StandingsLists[0].ConstructorStandings;
        return list.map((e) => ({ position: Number(e.position), team_id: e.Constructor.constructorId, points: Number(e.points) }));
      } catch (e) { /* fall through */ }
    }
    return payload.data || [];
  }

  async function init() {
    const [driverPayload, constructorPayload, driverMap] = await Promise.all([
      F1.api("/standings/api/driver-standings"),
      F1.api("/standings/api/constructor-standings"),
      F1.getDriverMap(),
    ]);

    const drivers = normalizeDrivers(driverPayload);
    const constructors = normalizeConstructors(constructorPayload);

    const sourceLabel = { live: "Live · Jolpica API", local: "Season snapshot", simulated: "Simulated fallback" }[driverPayload.source] || driverPayload.source || "";
    $("#source-badge").textContent = sourceLabel;

    renderDriverTable(drivers, driverMap);
    renderConstructorTable(constructors);
    renderDriverChart(drivers, driverMap);
    renderShareChart(constructors);
  }

  function teamName(id) {
    const names = { mclaren: "McLaren", ferrari: "Ferrari", redbull: "Red Bull Racing", mercedes: "Mercedes", astonmartin: "Aston Martin", williams: "Williams", audi: "Audi", alpine: "Alpine", haas: "Haas", racingbulls: "Racing Bulls", cadillac: "Cadillac" };
    return names[id] || id;
  }

  function renderDriverTable(drivers, driverMap) {
    const head = `<div class="fs-10 uppercase tracking-widest font-bold px-4 py-2 surface-alt text-sub" style="display:grid;grid-template-columns:36px 1fr 90px 120px;gap:8px">
      <span>#</span><span>Driver</span><span>Points</span><span>Team</span>
    </div>`;
    const rows = drivers.map((d) => {
      const info = driverMap[d.driver_code] || {};
      const color = info.team_color || F1.teamColor(d.team);
      return `<div style="display:grid;grid-template-columns:36px 1fr 90px 120px;gap:8px;border-top:1px solid var(--border)" class="items-center px-4 py-2.5 text-sm">
        <span class="f1-mono font-bold">${d.position}</span>
        <span class="flex items-center gap-2 min-w-0">
          <span class="team-bar" style="background:${color}"></span>
          <span class="min-w-0">
            <div class="f1-display font-bold truncate">${F1.escapeHtml(info.name || d.driver_code)}</div>
            <div class="fs-10 text-sub truncate">${F1.escapeHtml(info.team_name || teamName(d.team))}</div>
          </span>
        </span>
        <span class="f1-mono font-bold">${d.points}</span>
        <span class="fs-11 text-sub">${F1.escapeHtml(info.team_name || teamName(d.team))}</span>
      </div>`;
    }).join("");
    $("#driver-table-wrap").innerHTML = `<div class="f1-table" style="min-width:520px">${head}${rows}</div>`;
  }

  function renderConstructorTable(constructors) {
    const head = `<div class="fs-10 uppercase tracking-widest font-bold px-4 py-2 surface-alt text-sub" style="display:grid;grid-template-columns:36px 1fr 90px;gap:8px">
      <span>#</span><span>Constructor</span><span>Points</span>
    </div>`;
    const rows = constructors.map((t) => `<div style="display:grid;grid-template-columns:36px 1fr 90px;gap:8px;border-top:1px solid var(--border)" class="items-center px-4 py-2.5 text-sm">
      <span class="f1-mono font-bold">${t.position}</span>
      <span class="flex items-center gap-2">
        <span class="team-bar" style="background:${F1.teamColor(t.team_id)}"></span>
        <span class="f1-display font-bold">${F1.escapeHtml(teamName(t.team_id))}</span>
      </span>
      <span class="f1-mono font-bold">${t.points}</span>
    </div>`).join("");
    $("#constructor-table-wrap").innerHTML = `<div class="f1-table" style="min-width:420px">${head}${rows}</div>`;
  }

  function renderDriverChart(drivers, driverMap) {
    const top = drivers.slice(0, 12);
    F1Charts.barDistribution(
      $("#chart-driver-points"),
      top.map((d) => d.driver_code),
      top.map((d) => d.points),
      top.map((d) => (driverMap[d.driver_code] || {}).team_color || F1.teamColor(d.team)),
      {}
    );
  }

  function renderShareChart(constructors) {
    F1Charts.doughnut(
      $("#chart-points-share"),
      constructors.map((t) => teamName(t.team_id)),
      constructors.map((t) => t.points),
      constructors.map((t) => F1.teamColor(t.team_id))
    );
  }

  document.addEventListener("DOMContentLoaded", init);
})();
